tests/e2e/k8s/util/k8s_yaml.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2023 The ClusterLink Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/clusterlink-net/clusterlink/pkg/bootstrap/platform"
+)
+
+// replaceExact replaces <search> exactly <count> times.
+func replaceExact(s, search, replace string, count int) (string, error) {
+	searchCount := strings.Count(s, search)
+	if searchCount != count {
+		return "", fmt.Errorf("found %d (!=%d) occurrences of '%s'", searchCount, count, search)
+	}
+
+	return strings.ReplaceAll(s, search, replace), nil
+}
+
+// remove removes a substring starting with <from> until <to> (excluding).
+func remove(s, from, to string) (string, error) {
+	searchCount := strings.Count(s, from)
+	if searchCount != 1 {
+		return "", fmt.Errorf("found %d (!=1) occurrences of '%s'", searchCount, from)
+	}
+
+	startPos := strings.Index(s, from)
+	tmpPos := strings.Index(s[startPos+len(from):], to)
+	if tmpPos == -1 {
+		return "", fmt.Errorf("cannot found termination for '%s'", from)
+	}
+	endPos := startPos + len(from) + tmpPos
+
+	return s[:startPos] + s[endPos:], nil
+}
+
+func (f *Fabric) generateK8SYAML(p *peer, cfg *PeerConfig) (string, error) {
```

Comment:
`pkg/bootstrap` generates a YAML, and I want the suite to use that code.
So to align to your thoughts, we need to change `pkg/bootstrap` to base on k8s go objects, instead of a yaml template.
I think the template is more readable.
I agree that on the test side it's more "magic" (or "hacky"), but it's just the test code.

Reply:
so a recommendation may be to split each text replacement to its own function. It will make the code a more readable as the current indentation seems quite different than what you'd expect from the rest of the code base